src/imageBuild/buildToolsRegistry.ts

```diff
@@ -0,0 +1,94 @@
+import * as vscode from 'vscode';
+
+import { getPreferBuildah } from '../components/config/config';
+import { shell } from '../shell';
+
+interface ContainerImageBuildTool {
+
+    /** Returns a command line to build the specified image. */
+    getBuildCommand(image: string): string;
+
+    /** Returns a command line to push the specified image. */
+    getPushCommand(image: string): string;
+
+    /** Tests whether this tool is installed and properly configured. */
+    available(): Promise<boolean>;
+}
+
+class Docker implements ContainerImageBuildTool {
+    private binName = 'docker';
+
+    getBuildCommand(image: string): string {
+        return `${this.binName} build -t ${image} .`;
+    }
+
+    getPushCommand(image: string): string {
+        return `${this.binName} push ${image}`;
+    }
+
+    async available(): Promise<boolean> {
+        const result = await shell.exec(`${this.binName} version --format "{{.Server.APIVersion}}"`);
+        return result !== undefined && result.code === 0;
+    }
+}
+
+class Buildah implements ContainerImageBuildTool {
+    private binName = 'buildah';
+
+    getBuildCommand(image: string): string {
+        return `${this.binName} bud -t ${image} .`;
+    }
+
+    getPushCommand(image: string): string {
+        return `${this.binName} push ${image}`;
+    }
+
+    async available(): Promise<boolean> {
+        const result = await shell.exec(`${this.binName} version`);
+        return result !== undefined && result.code === 0;
+    }
+}
+
+const supportedTools: { [id: string]: ContainerImageBuildTool } = {
+    Docker: new Docker(),
+    Buildah: new Buildah()
+};
+
+/** Returns a command line for building the given image. */
+export function getBuildCommand(image: string): string {
+    const buildTool = supportedTools[currentBuildTool];
+    return buildTool.getBuildCommand(image);
+}
+
+/** Returns a command line for pushing the given image. */
+export function getPushCommand(image: string): string {
+    const buildTool = supportedTools[currentBuildTool];
+    return buildTool.getPushCommand(image);
+}
+
+/** Currently used image build tool. */
+export let currentBuildTool: string;
+
+/** Starts listening changing the preferred build tool. */
+export function start(): vscode.Disposable {

```

Comment:
What would be the cost of just looking up the config entry each time, instead of having a mutable variable and an event listener?  We do this throughout the extension and it feels simpler and doesn't seem to cause performance problems (he said cautiously, acutely aware that we _do_ have performance problems).

Reply:
Oh, it's because it's not just looking in config, it's also checking if `buildah` is present.  And we definitely _don't_ want to do that every time!

Reply:
For most tools we defer detection until the user tries to use the tool.  I think we check for `kubectl` on startup but that is just to warn - we don't rely on it for a setting.  We also have existing mechanisms for checking if a binary is present (see e.g. `binutil.ts`) which might be faster than calling the executable.